lib/adapters/call-hierarchy-adapter.ts

```diff
@@ -0,0 +1,145 @@
+import type * as atomIde from "atom-ide-base"
+import Convert from "../convert"
+import * as Utils from "../utils"
+import { SymbolTag } from "../languageclient"
+import type { LanguageClientConnection, ServerCapabilities, CallHierarchyItem } from "../languageclient"
+import type { CancellationTokenSource } from "vscode-jsonrpc"
+import type { Point, TextEditor } from "atom"
+
+import OutlineViewAdapter from "./outline-view-adapter"
+
+/** Public: Adapts the documentSymbolProvider of the language server to the Outline View supplied by Atom IDE UI. */
+export default class CallHierarchyAdapter {
+  private _cancellationTokens: WeakMap<LanguageClientConnection, CancellationTokenSource> = new WeakMap()
+
+  /**
+   * Public: Determine whether this adapter can be used to adapt a language server based on the serverCapabilities
+   * matrix containing a callHierarchyProvider.
+   *
+   * @param serverCapabilities The {ServerCapabilities} of the language server to consider.
+   * @returns A {Boolean} indicating adapter can adapt the server based on the given serverCapabilities.
+   */
+  public static canAdapt(serverCapabilities: ServerCapabilities): boolean {
+    return !!serverCapabilities.callHierarchyProvider
+  }
+
+  /** Corresponds to lsp's CallHierarchyPrepareRequest */
+
+  /**
+   * Public: Obtain the relationship between calling and called functions hierarchically. Corresponds to lsp's
+   * CallHierarchyPrepareRequest.
+   *
+   * @param connection A {LanguageClientConnection} to the language server that provides highlights.
+   * @param editor The Atom {TextEditor} containing the text associated with the calling.
+   * @param position The Atom {Point} associated with the calling.
+   * @param type The hierarchy type either incoming or outgoing.
+   * @returns A {Promise} of an {CallHierarchy}.
+   */
+  async getCallHierarchy<T extends atomIde.CallHierarchyType>(
+    connection: LanguageClientConnection,
+    editor: TextEditor,
+    point: Point,
+    type: T
+  ): Promise<atomIde.CallHierarchy<T>> {
+    const results = await Utils.doWithCancellationToken(connection, this._cancellationTokens, (cancellationToken) =>
+      connection.prepareCallHierarchy(
+        {
+          textDocument: Convert.editorToTextDocumentIdentifier(editor),
+          position: Convert.pointToPosition(point),
+        },
+        cancellationToken
+      )
+    )
+    return <CallHierarchyForAdapter<T>>{
+      type,
+      data: results?.map(convertCallHierarchyItem) ?? [],
+      itemAt(n: number) {
```

Comment:
Using a more descriptive name maybe?
```suggestion
      itemAt(num: number) {
```